contracts/tge/HATAirdrop.sol

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.16;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/cryptography/MerkleProofUpgradeable.sol";
+import "../tokenlock/ITokenLockFactory.sol";
+
+/*
+An airdrop contract that transfers tokens based on a merkle tree.
+*/
+contract HATAirdrop is OwnableUpgradeable {
+    error CannotRedeemBeforeStartTime();
+    error CannotRedeemAfterDeadline();
+    error LeafAlreadyRedeemed();
+    error InvalidMerkleProof();
+    error CannotRecoverBeforeDeadline();
+
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    bytes32 public root;
+    uint256 public startTime;
+    uint256 public deadline;
+    uint256 public lockEndTime;
+    uint256 public periods;
```

Comment:
this can be calculated param(once a day)

Reply:
But to calculate this would mean we need to hardcode a period length (or pass it and then it's just the same no?)

Reply:
So I think it's more flexible just directly passing periods count

Reply:
I don't get what you are saying Ben. We know the duration (lockEndTime - startTime). If we want to hardcode "once a day" as shay suggests, we can calculate periods as `(lockEndTime - startTime)/24h`

Reply:
Yes that's what I'm saying that we'd need to hardcode it, but why would we want to hardcode a value we are not sure about? It's better to leave periods completely flexible by letting the deployer pass it